### PR TITLE
Framework agnostic softmax kernels

### PR DESCRIPTION
**Pull request by ksivaman**
This PR separates the code and compilation of softmax kernels from the framework. Separate cuda extensions in PyTorch are added that use these kernels for PyTorch TE.

### CODE REVIEW

**Comment by ksivaman** [comment]
Tested for perf to ensure no regressions from baseline @ptrendx

File: setup.py [hunk 1]
```diff
@@ -267,7 +198,7 @@ def build_extensions(self) -> None:
             "-DCMAKE_LIBRARY_OUTPUT_DIRECTORY_{}={}".format(config.upper(), build_dir),
         ]
 
-        cmake_build_args = ["--config", config]
+        cmake_build_args = ["--config", config, "-v"]
```
Comment (ptrendx) [inline setup.py:201]: I think that does not need to be there, right?

File: transformer_engine/common/CMakeLists.txt [hunk 2]
```diff
@@ -5,7 +5,7 @@
 cmake_minimum_required(VERSION 3.18)
 
 if(NOT DEFINED CMAKE_CUDA_ARCHITECTURES)
-  set(CMAKE_CUDA_ARCHITECTURES 70 80 90)
+  set(CMAKE_CUDA_ARCHITECTURES 70 75 80 90)
```
Comment (ptrendx) [inline transformer_engine/common/CMakeLists.txt:8]: Turing can be handled by sm_70 and having special versions for it will just increase the library size - can we remove that?

File: transformer_engine/common/CMakeLists.txt [hunk 3]
```diff
@@ -30,7 +30,9 @@ add_library(transformer_engine SHARED
                                layer_norm/ln_api.cpp
                                layer_norm/ln_bwd_semi_cuda_kernel.cu
                                layer_norm/ln_fwd_cuda_kernel.cu
-                               util/cast.cu)
+                               util/cast.cu
+                               fused_softmax/scaled_masked_softmax_cuda.cu
+                               fused_softmax/scaled_upper_triang_masked_softmax_cuda.cu)
```
Comment (ptrendx) [inline transformer_engine/common/CMakeLists.txt:35]: I would probably drop the `_cuda` from the name, considering it is `.cu` file already :-)

File: transformer_engine/common/fused_softmax/scaled_masked_softmax_cuda.cu [hunk 4]
```diff
@@ -1,45 +1,51 @@
 /*************************************************************************
- * Copyright (c) 2022, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+ * Copyright (c) 2022, NVIDIA CORPORATION. All rights reserved.
```
Comment (ptrendx) [inline transformer_engine/common/fused_softmax/scaled_masked_softmax_cuda.cu:2]: `& AFFILIATES` should be there.

File: transformer_engine/common/fused_softmax/scaled_masked_softmax_cuda.cu [hunk 5]
```diff
@@ -1,45 +1,51 @@
...
  *
- * See LICENSE for license information.
+ * See LICENSE.txt for license information
```
Comment (ptrendx) [inline transformer_engine/common/fused_softmax/scaled_masked_softmax_cuda.cu:4]: Same here, the old version is right.

File: transformer_engine/common/fused_softmax/scaled_masked_softmax_cuda.cu [hunk 6]
```diff
@@ -305,6 +305,13 @@ __global__ void scaled_masked_softmax_warp_forward(
     }
     warp_reduce<acc_t, WARP_BATCH, WARP_SIZE, Max>(max_value);
 
+    // compute scale value to account for full mask
+    acc_t scale_value[WARP_BATCH];
```
Comment (ptrendx) [inline transformer_engine/common/fused_softmax/scaled_masked_softmax_cuda.cu:309]: Neither this nor the max_value is assigned the default value. You need to be REALLY careful that the value gets assigned for all elements that are going to be read from (even for inputs that are potentially invalid), as otherwise this is an undefined behavior (and compiler is free to do anything in that case.

File: transformer_engine/common/fused_softmax/scaled_masked_softmax_cuda.cu [hunk 7]
```diff
@@ -305,6 +305,13 @@ __global__ void scaled_masked_softmax_warp_forward(
...
+#pragma unroll
+    for (int i = 0;  i < WARP_BATCH;  ++i) {
+        scale_value[i] = (max_value[i] == -10000.0) ? 0.0 : 1.0;
```
Comment (ptrendx) [inline transformer_engine/common/fused_softmax/scaled_masked_softmax_cuda.cu:312]: `-10000.0` seems like a very specific number, maybe at least do `<=`?

File: transformer_engine/common/fused_softmax/scaled_masked_softmax_cuda.cu [hunk 8]
```diff
@@ -610,8 +619,9 @@ void dispatch_scaled_masked_softmax_forward(
     int key_seq_len,
     int batches,
     int attn_heads,
-    int pad_batches) {
-    TORCH_INTERNAL_ASSERT(key_seq_len >= 0 && key_seq_len <= 4096);
+    int pad_batches,
+    cudaStream_t stream) {
+    NVTE_CHECK(key_seq_len >= 0 && key_seq_len <= 4096);
```
Comment (ptrendx) [inline transformer_engine/common/fused_softmax/scaled_masked_softmax_cuda.cu:624]: An error message would be nice.

File: transformer_engine/common/fused_softmax/scaled_masked_softmax_cuda.cu [hunk 9]
```diff
@@ -632,140 +643,140 @@ void dispatch_scaled_masked_softmax_forward(
 
         int warps_per_block = (threads_per_block / warp_size);
         int batches_per_block = warps_per_block * batches_per_warp;
-        TORCH_INTERNAL_ASSERT(query_seq_len%batches_per_block == 0);
+        NVTE_CHECK(query_seq_len%batches_per_block == 0);
```
Comment (ptrendx) [inline transformer_engine/common/fused_softmax/scaled_masked_softmax_cuda.cu:646]: Same, an error message would be very nice.

File: transformer_engine/common/fused_softmax/scaled_masked_softmax_cuda.cu [hunk 10]
```diff
@@ -782,8 +793,9 @@ void dispatch_scaled_masked_softmax_backward(
     int query_seq_len,
     int key_seq_len,
     int batches,
-    int attn_heads) {
-    TORCH_INTERNAL_ASSERT(key_seq_len >= 0 && key_seq_len <= 4096);
+    int attn_heads,
+    cudaStream_t stream) {
+    NVTE_CHECK(key_seq_len >= 0 && key_seq_len <= 4096);
```
Comment (ptrendx) [inline transformer_engine/common/fused_softmax/scaled_masked_softmax_cuda.cu:798]: Same.

File: transformer_engine/common/fused_softmax/scaled_masked_softmax_cuda.cu [hunk 11]
```diff
@@ -944,4 +957,178 @@ void dispatch_scaled_masked_softmax_backward(
...
+    Tensor *softmax_results,
+    float scale_factor,
+    cudaStream_t stream) {
+
+    const int batches = input.shape[0];
+    const int attn_heads = input.shape[1];
+    const int query_seq_len = input.shape[2];
+    const int key_seq_len = input.shape[3];
+
+    TRANSFORMER_ENGINE_TYPE_SWITCH_16BIT(input.dtype, softmax_type,
```
Comment (ptrendx) [inline transformer_engine/common/fused_softmax/scaled_masked_softmax_cuda.cu:972]: Does this softmax only work for 16-bit values? That is some pretty serious limitation :-(.

File: transformer_engine/common/fused_softmax/scaled_masked_softmax_cuda.cu [hunk 12]
```diff
@@ -944,4 +957,178 @@ void dispatch_scaled_masked_softmax_backward(
...
+            batches,
+            attn_heads,
+            stream););
+}
+
+
+}  // end namespace transformer_engine
+
+
+int get_batch_per_block_cuda(int query_seq_len, int key_seq_len, int batches, int attn_heads) {
```
Comment (ptrendx) [inline transformer_engine/common/fused_softmax/scaled_masked_softmax_cuda.cu:1069]: What is with all those `_cuda` names?

Comment (ptrendx) [reply transformer_engine/common/fused_softmax/scaled_masked_softmax_cuda.cu:1069]: Also, those should probably live in some anonymous namespace or something.

File: transformer_engine/common/fused_softmax/scaled_masked_softmax_cuda.cu [hunk 13]
```diff
@@ -944,4 +957,178 @@ void dispatch_scaled_masked_softmax_backward(
...
+    using namespace transformer_engine;
+    return get_batch_per_block(query_seq_len, key_seq_len, batches, attn_heads);
+}
+
+
+void nvte_scaled_softmax_forward_cuda(
```
Comment (ptrendx) [inline transformer_engine/common/fused_softmax/scaled_masked_softmax_cuda.cu:1075]: Again `_cuda` - this time in the actual user-facing API. Not nice.

File: transformer_engine/common/include/transformer_engine/softmax.h [hunk 14]
```diff
@@ -0,0 +1,82 @@
...
+#include <cuda_fp16.h>
+#include <cuda_bf16.h>
+#include "transformer_engine.h"
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+
+inline int log2_ceil(int value) {
```
Comment (ptrendx) [inline transformer_engine/common/include/transformer_engine/softmax.h:19]: This is an actual API header that exposes library functions. This random function should not be here. Please move it to some file like utils or common that is internal to TE.

File: transformer_engine/common/include/transformer_engine/softmax.h [hunk 15]
```diff
@@ -0,0 +1,82 @@
...
+
+void nvte_scaled_softmax_backward_cuda(
+    const NVTETensor output_grads,
+    const NVTETensor softmax_results,
+    float scale_factor,
+    cudaStream_t stream
+);
+
+
+int get_batch_per_block_cuda(int query_seq_len, int key_seq_len, int batches, int attn_heads);
```
Comment (ptrendx) [inline transformer_engine/common/include/transformer_engine/softmax.h:42]: Same.

File: transformer_engine/common/include/transformer_engine/softmax.h [hunk 16]
```diff
@@ -0,0 +1,82 @@
...
+
+void nvte_scaled_masked_softmax_backward_cuda(
+    const NVTETensor input,
+    NVTETensor softmax_results,
+    float scale_factor,
+    cudaStream_t stream
+);
+
+
+void nvte_scaled_upper_triang_masked_softmax_forward_cuda(
```
Comment (ptrendx) [inline transformer_engine/common/include/transformer_engine/softmax.h:62]: Are those upper triang versions important enough that they need to have their own API entrypoints? Maybe we can just have an option to the masked_softmax api call?

File: transformer_engine/pytorch/csrc/extensions.cu [hunk 17]
```diff
@@ -393,8 +393,226 @@ at::Tensor cast_from_fp8(const at::Tensor &input,
...
 
 
+at::Tensor scaled_softmax_forward(at::Tensor input,
+                                  float scale_factor
+) {
+    using namespace transformer_engine;
+    AT_ASSERTM(input.dim() == 4, "expected 4D tensor");
+    AT_ASSERTM((input.scalar_type() == at::ScalarType::Half) ||
+               (input.scalar_type() == at::ScalarType::BFloat16),
+               "Only fp16 and bf16 are supported");
```
Comment (ptrendx) [inline transformer_engine/pytorch/csrc/extensions.cu:403]: This sucks - we should have some plan to lift that limitation at least in the followup PR.

File: transformer_engine/pytorch/softmax.py [hunk 18]
```diff
@@ -222,6 +212,4 @@ def forward_torch_softmax(
     @staticmethod
     def get_batch_per_block(sq: int, sk: int, b: int, np: int) -> int:
```
Comment (ptrendx) [inline transformer_engine/pytorch/softmax.py:213]: Why do I as a user need to think about this thing? That should be hidden from me!

**Review by ptrendx — CHANGES_REQUESTED** [review]
Left a whole bunch of comments, not everything marked (like not every NVTE_CHECK call), but the comments apply to the not-marked sections too :-)

File: transformer_engine/common/fused_softmax/scaled_masked_softmax_cuda.cu (hunk 7 above)
Comment (ksivaman) [reply transformer_engine/common/fused_softmax/scaled_masked_softmax_cuda.cu:312]: This is set previously in line 284 and we would presumably want the exact value.

File: transformer_engine/common/fused_softmax/scaled_masked_softmax_cuda.cu (hunk 6 above)
Comment (ksivaman) [reply transformer_engine/common/fused_softmax/scaled_masked_softmax_cuda.cu:309]: Immediately after declaring `max_value` and `scale_value`, all elements in the array upto `WARP_BATCH` are populated, so this should be fine?

Comment (ptrendx) [reply transformer_engine/common/fused_softmax/scaled_masked_softmax_cuda.cu:309]: Ok, I see that :+1:

File: transformer_engine/common/fused_softmax/scaled_masked_softmax_cuda.cu (hunk 11 above)
Comment (ksivaman) [reply transformer_engine/common/fused_softmax/scaled_masked_softmax_cuda.cu:972]: Yes, currently this restriction is present. For `FP32`, the framework softmax is used in TE `softmax.py`.

File: transformer_engine/pytorch/softmax.py (hunk 18 above)
Comment (ksivaman) [reply transformer_engine/pytorch/softmax.py:213]: This is in `softmax.py`, which is not directly user facing. It is required to figure out if the fused softmax kernel for the given configuration (batch size, sequence length etc.) is available or if the framework version should be used.

File: transformer_engine/common/include/transformer_engine/softmax.h (hunk 16 above)
Comment (ksivaman) [reply transformer_engine/common/include/transformer_engine/softmax.h:62]: This scaled upper triangular masked softmax is the default path for most (all?) GPT3 model sizes we test

File: transformer_engine/common/fused_softmax/scaled_masked_softmax_cuda.cu (hunk 12 above)
Comment (ksivaman) [reply transformer_engine/common/fused_softmax/scaled_masked_softmax_cuda.cu:1069]: Re: anonymous namespace; we need this visible for the extension

Comment (ksivaman) [reply transformer_engine/common/fused_softmax/scaled_masked_softmax_cuda.cu:1069]: We could remove this function from the cpp side and the extensions entirely (currently it is also exposed in the API header `softmax.h`) and move it to the python side in `softmax.py`.

Comment (ptrendx) [reply transformer_engine/common/fused_softmax/scaled_masked_softmax_cuda.cu]: Both things are bad, but let's move it to Python side - it is better than having this random internal function in the C API. Going forward we need to make sure that this softmax API supports all shapes.

**Review by ptrendx — APPROVED** [review]
Thanks for answering all the comments.
